### DISCUSSION TITLE
Compile transitive dependency issues

### DISCUSSION THREAD

**Discussion by FlyingCat2023** [ginkgo-project/ginkgo#1554] [Q&A]
Hi all,
First of all, I created a shared library lib_A, linking to Ginkgo. In library lib_A, I have a class in which func_a using following codes to generate a Csr-type matrix.
    auto A = gko::share(mtx::create(exec)); 
    A.get()->read(matrix_assembly_data);
There is no issue when i compile all codes and add them to lib_A.
Secondly, I created another shared libary lib_B. Libray lib_B is linked to lib_A. In somewhere Lib_B calls the func_a in lib_A. Then compilation failed to generate the lib_B, Error as below.

/usr/bin/ld: CMakeFiles/test_ginkgo.dir/test_ginkgo.cpp.o:(.data.rel.ro._ZTVN3gko6matrix3CsrIdmEE[_ZTVN3gko6matrix3CsrIdmEE]+0x80): undefined reference to `gko::matrix::Csr<double, unsigned long>::apply_impl(gko::LinOp const*, gko::LinOp const*, gko::LinOp const*, gko::LinOp*) const'

**Comment by upsj** [comment]
You are trying to use an unsigned `IndexType`, Ginkgo only supports signed index types

**Reply by upsj** [reply]
I opened an issue #1555 to make such errors visible before the linking stage

**Reply by FlyingCat2023** [reply]
Hi Upsj,

Many thanks for the prompt reply. 
Indeed, the compilation error comes from the un-supported IndexType. Changing it back to singed type index solves this issue.
